R/reporter-junit.R

```diff
@@ -0,0 +1,150 @@
+    tests    = NULL,
+    root     = NULL,
+    suite    = NULL,
+    suite_time = NULL,
+
+    initialize = function(file = stdout()) {
+      super$initialize()
+      self$file    <- file
+    },
+
+    elapsed_time = function() {
+      time <- round((private$proctime() - self$timer)[["elapsed"]], 2)
+      self$timer <- private$proctime()
+      time
+    },
+
+    reset_suite = function () {
+      self$errors   <- 0
+      self$failures <- 0
+      self$skipped  <- 0
+      self$tests    <- 0
+      self$suite_time <- 0
+    },
+
+    start_reporter = function() {
+      self$timer <- private$proctime()
+      self$doc   <- xml2::xml_new_document()
+      self$root  <- xml2::xml_add_child(self$doc, 'testsuites')
+      self$reset_suite()
+    },
+
+    start_context = function(context) {
+      self$suite <- xml2::xml_add_child(self$root,
+        "testsuite",
+        name      = context,
+        timestamp = private$timestamp(),
+        hostname  = private$hostname()
+      )
+    },
+
+    end_context = function(context) {
+      xml2::xml_attr(self$suite, "tests") <- as.character(self$tests)
+      xml2::xml_attr(self$suite, "skipped") <- as.character(self$skipped)
+      xml2::xml_attr(self$suite, "failures") <- as.character(self$failures)
+      xml2::xml_attr(self$suite, "errors") <- as.character(self$errors)
+      xml2::xml_attr(self$suite, "time") <- as.character(self$suite_time)
+
+      self$reset_suite()
+    },
+
+    add_result = function(context, test, result) {
+      self$tests <- self$tests + 1
+
+      time <- self$elapsed_time()
+      self$suite_time <- self$suite_time + time
+
+      # XML node for test case
+      name <- test %||% "(unnamed)"
+      testcase <- xml2::xml_add_child(self$suite, "testcase",
+       time = toString(time),
+       classname = paste0(classnameOK(context), '.', classnameOK(name))
+      )
+
+      # message - if failure or error
+      message <- if (is.null(result$call)) "(unexpected)" else format(result$call)[1]
+
+      if (!is.null(result$srcref)) {
+        location <- paste0('@', attr(result$srcref, 'srcfile')$filename, '#', result$srcref[1])
+        message  <- paste(as.character(result), location)
+      }
+
+      # add an extra XML child node if not a success
+      if (expectation_error(result)) {
+        # "type" in Java is the exception class
+        xml2::xml_add_child(testcase, 'error', type = 'error', message = message)
+        self$errors <- self$errors + 1
+      } else if (expectation_failure(result)) {
+        # "type" in Java is the type of assertion that failed
+        xml2::xml_add_child(testcase, 'failure', type = 'failure', message = message)
+        self$failures <- self$failures + 1
+      } else if (expectation_skip(result)) {
+        xml2::xml_add_child(testcase, "skipped")
+        self$skipped <- self$skipped + 1
+      }
+    },
+
+    end_reporter = function() {
+      if (inherits(self$file, "connection")) {
+        file <- tempfile()
```

Comment:
Would you mind filing an issue for this on xml2 as well?  (It's fine to leave this workaround as is, but long-term xml2 should also supporting writing to connections)